Fix

lplewa opened the pull request:
### Description 
_Add a comprehensive description of proposed changes_


Fixes # - _issue number(s) if exists_

- [ ] - git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/oneapi-src/oneTBB/blob/master/CONTRIBUTING.md#pull-requests) for details)_

### Type of change

_Choose one or multiple, leave empty if none of the other choices apply_

_Add a respective label(s) to PR if you have permissions_

- [ ] bug fix - _change that fixes an issue_
- [ ] new feature - _change that adds functionality_
- [ ] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [ ] documentation - _documentation update_

### Tests

- [ ] added - _required for new features and some bug fixes_
- [ ] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [ ] needs to be updated
- [ ] not needed

### Breaks backward compatibility
- [ ] Yes
- [ ] No
- [ ] Unknown

### Notify the following users
_List users with `@` to send notifications_

### Other information

phprus committed:
Update macOS 10.15 to macOS 12 in CI (#1113)

Signed-off-by: Vladislav Shchapov <vladislav@shchapov.ru>

KFilipek committed:
Add issue labeler (#1070)

Signed-off-by: Krzysztof Filipek <krzysztof.filipek@intel.com>

Commit:
Fix build on systems without ucontext.h in libc (#1055)

* Enable thread based resumable tasks if ucontext is absent

Signed-off-by: pavelkumbrasev <pavel.kumbrasev@intel.com>
Co-authored-by: Ilya Isaev <ilya.isaev@intel.com>

brad0 committed:
Use CMAKE_DL_LIBS to properly choose to use libdl or not. (#1120)

Fixes build on *BSD with GCC. Same also done here for Clang..
https://github.com/oneapi-src/oneTBB/commit/dbccbee9f5527e802925852e225b1d1584f43a7a

pbalcer commented on src/tbbmalloc/backend.cpp:
```diff
@@ -813,7 +823,7 @@ FreeBlock *Backend::genericGetBlock(int num, size_t size, bool needAlignedBlock)
                                                         /*alignedBin=*/true, &numOfLockedBins);
             }
-        } while (!block && numOfLockedBins>lockedBinsThreshold);
+        } while (!block && (numOfLockedBins>lockedBinsThreshold || cleanCnt % 2 == 1 || cleanCnt != backendCleanCnt));
```
this looks like magic, at least add a comment.

pbalcer commented on src/tbbmalloc/large_objects.cpp:
```diff
@@ -1020,17 +1020,33 @@ void ExtMemoryPool::freeLargeObjectList(LargeMemoryBlock *head)
-bool ExtMemoryPool::hardCachesCleanup()
+bool ExtMemoryPool::hardCachesCleanup(bool wait)
 {
+    if (hardCachesCleanupInProgress.fetch_or(0x1)) {
```
this condition is the other way around compared to `softCachesCleanup`. I found this confusing since the only difference is `!` at the beginning. I'd suggest refactoring these to have the same flow.

lplewa committed:
Refactor waitTillBlockReleased function

"for loop" without condition and without iterator is not a great
solution for code readability. For simple infinite loop "while loop" is
much more readable solution.

Signed-off-by: Łukasz Plewa <lukasz.plewa@intel.com>

lplewa committed:
Add additional synchronization around *CachesCleanup function

This commit fixes the issue when test_malloc_pools TestFixedBufferPool
fails sporadically on test_malloc_pools.cpp:437

utils::NativeParallelFor( p, FixedPoolUse(p, pool, MAX_OBJECT/p/2,1) );

The main problem was that threads were not aware of cleanup operations
done by SoftCachesCleanup() and HardCachesCleanup() functions.
So while one thread was doing a cleanup, the second one
returns ENOMEM as it couldn't find any memory to allocate in the
backend (this test creates fixedSizePool)

The change adds additional flags and counters that indicate cleanup
operation is in progress, allowing the thread to wait for those
operations to complete before deciding if there is free memory to
allocate or not and return this information to the caller.

In the scenario reproduced by the test, memory for the allocation can
be reclaimed from the LOC. This operation happens in the
loc.cleanAll() or loc.regularCleanup functions called respectively, by
hardCachesCleanup() or softCachesCleanup() functions.
In those functions, free memory is removed from LOC and added to the
backend bin. Between those operations, other threads are unaware that
this memory block is free and returns ENOMEM to the user.

To fix this problem, this patch adds two flags to indicate that
cleanup (soft or hard) is in progress, and threads should wait before
returning NULL to the user.

But there was one more problem: backend::clean() function.
This function removes memory from the bins to try to release or
coalesce it. The problem is that other threads were not aware of
this operation. It causes a 'funny' situation when:

Thread 1: Performs HardCachescleanup function
Thread 2: wait "waitTillBlockReleased()"
Thread 1: Returns memory from LOC
Thread 2: exits waitTillBlockReleased() as it detected new memory in
backend
Thread 1: starts backend::clean() - removes memory from the bin to try
coalesce it
Thread 2: Fail memory allocation as there is no free block in
bins, and returns ENOMEM
Thread 1: Finish coalescing and return memory back to bins

To resolve this problem, this patch adds a counter which allows
threads to detect when clean operation is done so that the threads
will wait for the complition of there is no memory available.

Signed-off-by: Łukasz Plewa <lukasz.plewa@intel.com>
